### PR TITLE
Use Merkle Trees to speed up diff computations

### PR DESCRIPTION
**Pull request by cmc333333**
Currently, we're doing a lot of work when computing diffs. We convert both left and right trees into a dictionary indexed by node label. We then step through both, looking at each node for changes. That's a lot of wasted effort when the tree hasn't changed much and doesn't leverage the existing structure.

This changeset creates an immutable version of the node tree (which will hopefully creep into the rest of the app), which can be leveraged to cut off large swaths of computations before they occur. Parent nodes aggregate their children, so when two parent nodes are compared, we immediately know if we must compare their children, too. This results in massive speedups during diff computation.

It's been tested on FEC's 110 and CFPB's reg E. Reg Z is harder to verify, but in progress. In the two existing cases, the produced diffs contained more accurate values for the `child_labels` field, and the addition of `tagged_text: null` to several diff nodes (which should have no effect).

Additional changes:
* Split text-related diffing code from tree-related code
* A few of flake8's suggestions

### CODE REVIEW

**Comment by khandelwal** [comment]
Using the dictionary indexed by node label was an attempt to stop traversing the tree each time to find a node.

**Comment by khandelwal** [comment]
It made sense at the time.

File: generate_diff.py [hunk 1]
```diff
@@ -1,12 +0,0 @@
-from regparser.diff import api_reader
```
Comment (khandelwal) [inline generate_diff.py]: Hah, it's the little scripts like this one that kept me from running everything.

File: regparser/diff/tree.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    changes.extend(_local_changes(lhs, rhs))
+
+    # Removed children. Note params reversed
+    removed_children = _new_in_rhs(rhs.children, lhs.children)
+    changes.extend(map(_data_for_delete, removed_children))
+    # grandchildren which appear to be deleted, but may just have been moved
+    possibly_moved = _index_by_label_id(
+        grandchild
+        for child in removed_children for grandchild in child.children)
```
Comment (khandelwal) [inline regparser/diff/tree.py]: This line is concise, but hard to read. Can we rewrite to make it a bit more explicit and easier to read?

File: regparser/diff/tree.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def _data_for_delete(node):
+    return (node.label_id, {"op": DELETED})
+
+
+def changes_between(lhs, rhs):
+    """Main entry point for this library. Recursively return a list of changes
+    between the lhs and rhs. lhs and rhs should be FrozenNodes. Note that this
+    *does not* account for reordering nodes, though it does account for
```
Comment (khandelwal) [inline regparser/diff/tree.py]: Is tracking reordering something that needs to be implemented? Or do we just express that as a delete, and an add (without linking the two)? It's not clear from this comment that it's something that needs to be implemented in the future or whether the current state is acceptable.

**Comment by khandelwal** [comment]
Should we have more tests around FrozenNode? We're also adding some new methods that aren't unit-tested explicitly? Wondering if we need to do that.

File: regparser/diff/tree.py (hunk 3 above)
Comment (cmc333333) [inline regparser/diff/tree.py]: We currently don't implement reordering in diffs (with a few exceptions). I don't know if that's acceptable or not, just noted it.

**Comment by khandelwal** [comment]
@cmc333333 Ping!

**Comment by cmc333333** [comment]
Updated to account for feedback

**Pull request merged**